File: LeetCode/Easy/1_two_sum.ts
// the function take num: an arry of numbers | target = numbers | it return arry of numbers []

function twoSum(nums: number[], target: number): number[] {

     // create a map that will store the numbers and corresponding indices
     //new Map() creates an empty Map object, ready for use. | without it the variable would be undefined
    const hashmap: Map<number, number> = new Map();
    for (let i = 0; i < nums.length; i++) {

     //Example: If target = 9 and nums[i] = 2, then complement = 9 - 2 = 7.
        const complement = target - nums[i];
        //Checks if the complement is already in the Map.
        if (hashmap.has(complement)) {
          //If true, it means a previous number and the current number sum up to the target.
          //The ! indicates that the value is guaranteed to exist (TypeScript syntax).
            return [hashmap.get(complement)!, i];
        }
        //Add Current Number to the Map | Each number (nums[i]) is stored as a key, and its index (i) is stored as the value.
        hashmap.set(nums[i], i);
    }
    // Return an empty array if no solution is found
    return [];
}
 // Example usage
 const nums = [2, 7, 11, 15];
 const target = 9;
 const result = twoSum(nums, target);
 console.log("Indices of the two numbers:", result);

File: LeetCode/Easy/1_two_sum.py
class Solution:
    def twoSum(self, nums, target):
        # Dictionary to store numbers and their indices
        num_map = {}
        
        # Loop through the list, in this case i is the index num is the number 
        for i, num in enumerate(nums):
            # Calculate the complement
            complement = target - num
            
            # Check if the complement exists in the dictionary
            if complement in num_map:
                # If it exists, return the indices of the complement and the current number
                print('this is num_map', num_map[complement])
                return [num_map[complement], i]
            
            # If the complement doesn't exist, add the current number 
            # and its index to the dictionary
            num_map[num] = i
            print(num_map)
           
        
        # If no solution is found (though the problem guarantees one)
        return []
    
nums = [2, 7, 11, 15]
target = 9
# in python method that have self need an instance to call them so self refers to that instance 
# creating instance of the solution class 
s = Solution()
result = s.twoSum(nums, target)
print("Result:", result)  # Output: [0, 1]
